File: week1-c/credit/credit.c
#include <stdio.h>
#include <cs50.h>

// User defined functions
int calculateCheckSum(long number);
int checkDigits(long number);
int getFirstDigits(long number);

// Main code
int main(void)
{
    // Ask the user for a credit card number
    long number;

    do
    {
        number = get_long("Number: ");
    }
    while (number <= 0);

    // Calculate checksum
    int sum = calculateCheckSum(number);

    // Calculate the number of digits
    int dig = checkDigits(number);

    // Get the card first 2 digits
    int fDig = getFirstDigits(number);

    // Checks the card issuer
    if (sum % 10 == 0 && dig == 15 && (fDig == 34 || fDig == 37))
    {
        printf("AMEX\n");
    }
    else if (sum % 10 == 0 && dig == 16 && fDig >= 51 && fDig <= 55)
    {
        printf("MASTERCARD\n");
    }
    else if (sum % 10 == 0 && (dig == 16 || dig == 13) && fDig / 10 == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

}

// Fuction to calculate checkSum
int calculateCheckSum(long number)
{
    int nDigits = 0;
    int sum = 0;
    int digit, temp;

    while (number > 0)
    {
        // Extracts the last digit
        digit = number % 10;

        if (nDigits % 2 == 0)
        {
            sum += digit;
        }
        else
        {
            temp = digit * 2;

            // Sums the digits to the total
            while (temp > 0)
            {
                sum += temp % 10;
                temp /= 10;
            }
        }

        // Removes last digit and tracks current position
        number /= 10;
        nDigits++;
    }

    return sum;
}

// Function to check the number of digits
int checkDigits(long number)
{
    int digits = 0;

    while (number > 0)
    {
        number /= 10;
        digits++;
    }

    return digits;
}

// Function to get first 2 digits
int getFirstDigits(long number)
{
    while (number / 100 > 0)
    {
        number /= 10;
    }

    return number;
}

File: week6-python/credit/credit.py
from cs50 import get_string


def main():
    # Main code

    while True:
        # Ask the user for a credit card number
        number = get_string("Number: ")

        if int(number) > 0:
            break

    # Calculate checksum
    Sum = calculateCheckSum(number)

    # Calculate the number of digits
    Dig = len(number)

    # Get the card first 2 digits
    fDig = int(number[:2])

    if Sum % 10 == 0 and Dig == 15 and (fDig == 34 or fDig == 37):
        print("AMEX")

    elif Sum % 10 == 0 and Dig == 16 and fDig in range(51, 56):
        print("MASTERCARD")

    elif Sum % 10 == 0 and (Dig == 16 or Dig == 13) and str(fDig)[0] == '4':
        print("VISA")

    else:
        print("INVALID")


def calculateCheckSum(number):
    # Function to calculate checkSum
    Sum = 0

    # Gets only half of the digits
    for digit in number[::-1][1::2]:
        digit = str(int(digit) * 2)
        for char in digit:
            Sum += int(char)

    # Sums the other half of digits
    for digit in number[::-1][::2]:
        Sum += int(digit)

    return Sum


# Run the main code
if __name__ == "__main__":
    main()
